src/shared/Core/Trace2.cs

```diff
@@ -449,34 +461,122 @@ public abstract class Trace2Message
     [JsonProperty("line", Order = 6)]
     public int Line { get; set; }
 
+    [JsonProperty("depth", Order = 7)]
+    public int Depth { get; set; }
+
     public abstract string ToJson();
 
     public abstract string ToNormalString();
 
-    protected string BuildNormalString(string message)
+    public abstract string ToPerformanceString();
+
+    protected abstract string BuildPerformanceSpan();
+
+    protected string BuildNormalString()
     {
+        string message = GetEventMessage(Trace2FormatTarget.Normal);
+
         // The normal format uses local time rather than UTC time.
-        string time = Time.ToLocalTime().ToString("HH:mm:ss.ffffff");
+        string time = Time.ToLocalTime().ToString(NormalPerfTimeFormat);
+        string source = GetSource();
+
+        // Git's TRACE2 normal format is:
+        // [<time> SP <filename>:<line> SP+] <event-name> [[SP] <event-message>] LF
+        return $"{time} {source,-33} {Event.ToString().ToSnakeCase()} {message}";
+    }
+
+    protected string BuildPerformanceString()
+    {
+        string message = GetEventMessage(Trace2FormatTarget.Performance);
+
+        // The performance format uses local time rather than UTC time.
+        var time = Time.ToLocalTime().ToString(NormalPerfTimeFormat);
```

Comment:
Just double checking that `NormalPerfTimeFormat` is the correct format for the performance target?

Reply:
Yes! I prefixed `TimeFormat` with `NormalPerf` since the normal and performance format targets use local time while the event format target uses UTC time. Not necessarily my best naming job, though - we can change in future if something better comes to us.